dds/DCPS/security/SSL/SubjectName.cpp

```diff
@@ -145,28 +139,204 @@ namespace SSL {
+  bool Parser::accept(char c)
+  {
+    if (pos_ == in_.size()) {
+      return false;
+    }
+    if (in_[pos_] == c) {
+      ++pos_;
+      return true;
+    }
+    return false;
+  }
+
+  bool Parser::relative_distinguished_name(AVAVec& store)
+  {
+    if (!attribute_type_value(store)) {
+      return false;
+    }
+    while (accept('+')) {
+      if (!attribute_type_value(store)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool Parser::attribute_type_value(AVAVec& store)
+  {
+    std::string at, av;
+    if (!attribute_type(at) || !accept('=') || !attribute_value(av)) {
+      return false;
+    }
+    store.push_back(std::make_pair(at, av));
+    return true;
+  }
+
+  bool Parser::attribute_type(std::string& at)
+  {
+    size_t equal_pos = in_.find_first_of("=", pos_);
+    if (equal_pos == std::string::npos) {
+      return false;
+    }
+    at = in_.substr(pos_, equal_pos - pos_);
+
+    pos_ = equal_pos;
+    return validate_attribute_type(at);
+  }
+
+  bool Parser::is_alpha(char c) const
+  {
+    return ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z');
+  }
+
+  bool Parser::validate_attribute_type(const std::string& at) const
+  {
+    // Attibute type must start with a lower-case alphabet and followed by
+    // any number of alphabets (lower or upper-case) or digits or hyphens.
+    if (at.empty() || !is_alpha(at[0])) {
+      return false;
+    }
+    for (std::string::size_type i = 0; i < at.size(); ++i) {
+      const char c = at[i];
+      if (!is_alpha(c) && !('0' <= c && c <= '9') && c != '-') {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool Parser::attribute_value(std::string& av)
+  {
+    // The first comma or plus character that is not escaped marks the end
+    // of the current value. If not found, the value runs to the end of the input.
+    std::string::size_type start_pos = pos_;
+    bool got_value = false;
+    while (start_pos < in_.size() && !got_value) {
+      std::string::size_type delimiter_pos = in_.find_first_of(",+", start_pos);
+      if (delimiter_pos == std::string::npos) {
+        av = in_.substr(pos_, in_.size() - pos_);
+        pos_ = in_.size();
+        got_value = true;
+      } else {
+        long long tmp_pos = delimiter_pos - 1;
+        size_t esc_count = 0;
+        while (tmp_pos >= 0 && in_[tmp_pos--] == '\\') {
+          ++esc_count;
+        }
+        if (esc_count % 2 == 0) {
+          av = in_.substr(pos_, delimiter_pos - pos_);
+          pos_ = delimiter_pos;
+          got_value = true;
+        } else {
+          start_pos = delimiter_pos + 1;
+        }
+      }
+    }
+
+    if (!got_value) {
+      av = in_.substr(pos_, in_.size() - pos_);
+      pos_ = in_.size();
+      got_value = true;
+    }
+
+    unescape(av);
+    return validate_attribute_value(av);
+  }
+
+  void Parser::unescape(std::string& av) const
+  {
+    // Unescape the escaped characters from the attribute value.
+    replace_all(av, "\\\\", "\\");
```

Comment:
I think this line needs to be last.  Otherwise, you could be creating a new escape sequence.

Reply:
I think this is fine because it's unescaping. But I've moved it to the last to be safe.